### PR TITLE
lnrpc+lncli: fix sat per virtual byte

### LINKED ISSUE #4308: lncli openchannel --sat_per_byte
### Background

Openening a channel using --sat_per_byte leads to a approx. half the fee value in the transaction.
Setting it to 20 sat/byte creates a transaction with ~10.8 sat/byte or 5 sat/WU. The only way I get 20 is by using vbytes. Assuming 1 WU is 4 vbytes.
Am I missing something obvious?

### Your environment

* lnd 0.10.0-beta
* Linux raspberrypi 4.19.97-v7l+ #1294 SMP armv7l GNU/Linux
* bitcoind 0.19.0

### Steps to reproduce

lncli openchannel .... -sat_per_byte [sats]

### PR DESCRIPTION
**Pull request by yyforyongyu**
This PR fixes #4308. It's done in two steps,
- First, although it will be deprecated, the docs for all the fields `sat_per_byte` are updated.
- Second, deprecate the field `sat_per_byte` and replace it with a new field `sat_per_vbyte`.

### CODE REVIEW

File: lnrpc/rpc.proto [hunk 1]
```diff
@@ -845,8 +845,11 @@ message EstimateFeeResponse {
     // The total fee in satoshis.
     int64 fee_sat = 1;
 
+    // Deprecated. The fee rate in satoshi/vbyte.
+    int64 feerate_sat_per_byte = 2 [deprecated = true];
+
     // The fee rate in satoshi/vbyte.
-    int64 feerate_sat_per_byte = 2;
+    uint64 feerate_sat_per_vbyte = 3;
```
Comment (carlaKC) [inline lnrpc/rpc.proto:852]: Perhaps take this opportunity to standardize `sat_per_vbyte` across all endpoints? I don't think we need the `feerate_` prefix here.

File: lnrpc/rpc.proto [hunk 2]
```diff
@@ -859,7 +862,11 @@ message SendManyRequest {
 
     // A manual fee rate set in sat/vbyte that should be used when crafting the
     // transaction.
-    int64 sat_per_byte = 5;
+    uint64 sat_per_vbyte = 4;
```
Comment (carlaKC) [inline lnrpc/rpc.proto:865]: Strange (and convenient) that field 4 is missing, but looks like it was just an onversight in `9b1f21e2`.

File: lnrpc/rpc.proto [hunk 3]
```diff
@@ -1679,9 +1698,9 @@ message OpenChannelRequest {
     // confirmed by.
     int32 target_conf = 6;
 
-    // A manual fee rate set in sat/vbyte that should be used when crafting the
-    // funding transaction.
-    int64 sat_per_byte = 7;
+    // Deprecated. A manual fee rate set in sat/vbyte that should be used when
```
Comment (carlaKC) [inline lnrpc/rpc.proto:1701]: nit: for these comments, mention the new field `Deprecated, use sat_per_byte`?

File: lnrpc/walletrpc/walletkit.proto [hunk 4]
```diff
@@ -400,11 +400,12 @@ message PendingSweep {
     uint32 amount_sat = 3;
 
     /*
+    Deprecated.
```
Comment (carlaKC) [inline lnrpc/walletrpc/walletkit.proto:403]: nit: standardize deprecation notice on its own line vs inline with the rest of the comment?

File: rpcserver.go [hunk 5]
```diff
@@ -213,6 +213,45 @@ func stringInSlice(a string, slice []string) bool {
 	return false
 }
 
+// calculateFeeRate uses either satPerByte or satPerVByte, but not both, from a
+// request to calculate the fee rate. It provides compatibility for the
+// deprecated field, satPerByte. Once the field is safe to be removed, the
+// check can then be deleted.
+func calculateFeeRate(satPerByte, satPerVByte uint64, targetConf uint32,
```
Comment (carlaKC) [inline rpcserver.go:220]: niceee👍

**Review by carlaKC — APPROVED** [review]
Much-needed clarification, thanks for the PR! Only nits from me.

**Comment by Roasbeef** [comment]
ping @cfromknecht

File: rpcserver.go [hunk 6]
```diff
@@ -216,6 +216,45 @@ func stringInSlice(a string, slice []string) bool {
...
+
+	// We only allow using either the deprecated field or the new field.
+	if satPerByte != 0 && satPerVByte != 0 {
+		return feeRate, fmt.Errorf("either SatPerByte or " +
+			"SatPerVByte should be set, but not both")
+	}
+
+	// Default to satPerVByte, and overwrite it if satPerByte is set.
+	satPerKw := chainfee.SatPerKVByte(satPerVByte * 1000).FeePerKWeight()
+	if satPerByte != 0 {
```
Comment (Roasbeef) [inline rpcserver.go:236]: 👍

**Review by Roasbeef — APPROVED** [review]
LGTM 🛍

Just needs a rebase! 

We'll also want to ensure that in the release notes we point out the new flags as we'll end up breaking any bash scripts the use these arguments. If we want to soften the move over, we can retain the parsing of the old version (sat/byte) on the command line, similar to how we do for the set of RPCs.

**Comment by yyforyongyu** [comment]
>  If we want to soften the move over, we can retain the parsing of the old version (sat/byte) on the command line, similar to how we do for the set of RPCs.

I think this approach is better, since this PR does change several cli flags at once. The old flags are now marked as deprecated and hidden. It won't show up in the help menu, while it's still able to pass them, thus no bash scripts broken.

File: cmd/lncli/cmd_open_channel.go [hunk 7]
```diff
@@ -120,7 +120,12 @@ var openChannelCommand = cli.Command{
 				"used for fee estimation",
 		},
 		cli.Int64Flag{
-			Name: "sat_per_byte",
+			Name:   "sat_per_byte",
+			Usage:  "Deprecated, use sat_per_vbyte instead.",
+			Hidden: true,
```
Comment (Roasbeef) [inline cmd/lncli/cmd_open_channel.go:125]: Oooo, didn't know this even existed!

File: cmd/lncli/main.go [hunk 8]
```diff
@@ -230,6 +230,23 @@ func extractPathArgs(ctx *cli.Context) (string, string, error) {
 	return tlsCertPath, macPath, nil
 }
 
+// checkNotBothSet accepts two flag names, a and b, and checks that only flag a
+// or flag b can be set, but not both. It returns the name of the flag or an
+// error.
+func checkNotBothSet(ctx *cli.Context, a, b string) (string, error) {
```
Comment (Roasbeef) [inline cmd/lncli/main.go:236]: 👍

**Review by Roasbeef — APPROVED** [review]
re-LGTM 🍕


Just needs a rebase!

**Comment by Roasbeef** [comment]
Something else landed before this, needs another rebase...

**Comment by yyforyongyu** [comment]
> Something else landed before this, needs another rebase...

Done!

**Pull request merged**